extension/server/src/providers/codeActions.ts

```diff
@@ -33,45 +38,273 @@ export default async function genericCodeActionsProvider(params: CodeActionParam
+	for (const input of inputs) {
+		const type = getType(input.keyword);
+		const declaration = type === `struct` ? 'dcl-ds' : 'dcl-s';
+		declarations.push(`  ${declaration} ${input.name} ${prettyKeywords(input.keyword, true)};`);
+		if (type === `struct`) {
+			for (const subItem of input.subItems) {
+				initializations.push(`  ${input.name}.${subItem.name} = ${getDefaultValue(getType(subItem.keyword))};`)
 
-		return refactorAction;
+				const structPath = subItem.position.path;
+				if (workspaceFolder) {
+					const relativePath = asPosix(path.relative(workspaceFolder.uri, structPath));
+					if (!includes.includes(relativePath)) {
+						includes.push(relativePath);
+					}
+				}
+			}
+		} else {
+			initializations.push(`  ${input.name} = ${getDefaultValue(type)};`);
+		}
+	}
+
+	return {
+		declarations,
+		initializations,
+		includes
 	}
 }
 
-export function getSubroutineActions(document: TextDocument, docs: Cache, range: Range): CodeAction|undefined {
+function getReturns(workspaceFolder: WorkspaceFolder | undefined, structs: Declaration[], currentProcedure: Declaration, name: 'actual' | 'expected') {
+	const declarations: string[] = [];
+	const initializations: string[] = [];
+	const assertions: string[] = [];
+	const includes: string[] = [];
+
+	const type = getType(currentProcedure.keyword);
+	const declaration = type === `struct` ? 'dcl-ds' : 'dcl-s';
+	if (type === `struct`) {
+		declarations.push(`  ${declaration} ${name} likeDS(${currentProcedure.keyword['LIKE']});`);
+		const struct = structs.find(struct => struct.name === currentProcedure.keyword[`LIKE`]);
+		if (struct) {
+			for (const subItem of struct.subItems) {
+				const subItemType = getType(subItem.keyword);
+				const subItemDefaultValue = getDefaultValue(subItemType);
+				initializations.push(`  ${name}.${subItem.name} = ${subItemDefaultValue};`)
+
+				const structPath = subItem.position.path;
+				if (workspaceFolder) {
+					const relativePath = asPosix(path.relative(workspaceFolder.uri, structPath));
+					if (!includes.includes(relativePath)) {
+						includes.push(relativePath);
+					}
+				}
+
+				const assertion = getAssertion(subItemType);
+				if (assertion === `assert`) {
+					assertions.push(`  ${assertion}(expected.${subItem.name} = actual.${subItem.name} : '${subItem.name}');`);
+				} else {
+					assertions.push(`  ${assertion}(expected.${subItem.name} : actual.${subItem.name} : '${subItem.name}');`);
+				}
+			}
+		}
+	} else {
+		declarations.push(`  ${declaration} ${name} ${prettyKeywords(currentProcedure.keyword, true)};`);
+		initializations.push(`  ${name} = ${getDefaultValue(type)};`);
+
+		const assertion = getAssertion(type);
+		if (assertion === `assert`) {
+			assertions.push(`  ${assertion}(expected = actual);`);
+		} else {
+			assertions.push(`  ${assertion}(expected : actual);`);
+		}
+	}
+
+	return {
+		declarations,
+		initializations,
+		assertions,
+		includes
+	}
+}
+
+function getType(keywords: Keywords): string {
```

Comment:
Something nice to do here would be to have it return a new type like: `PrimitiveType`